File: Javascript/3.Ottobre/71.TipCalculator-20Ottobre.js
//todo Dato il prezzo del pasto e una percentuale di mancia personalizzata, restituisci un array con tre valori di mancia: 15%, 20% e l'importo personalizzato.

//! I prezzi saranno indicati nel formato: "$N.NN".
//! Le percentuali di mancia personalizzate saranno indicate nel formato: "25%".
//! Restituisci gli importi nello stesso formato "$N.NN", arrotondati a due cifre decimali.
//! Ad esempio, dato un prezzo del pasto di "$10,00" e un valore di mancia personalizzato del "25%", restituisci ["$1,50", "$2,00", "$2,50"].

//? Tempo di esecuzione:
//* 16 Minuti e 32 Secondi 

//?Codice scritto senza utilizzo di ChatGpt

function calculateTips(mealPrice, customTip) {

  let mealPriceNumber = Number(mealPrice.replace(/\$/g, ""))
  let customTipNumber = Number(customTip.replace(/\%/g, ""))
  let scount15 = (mealPriceNumber * 0.15).toFixed(2).toString();
  let scount20 = (mealPriceNumber * 0.20).toFixed(2).toString();
  let scountCu = (mealPriceNumber * customTipNumber / 100).toFixed(2).toString();
  let finalWord = ["$" + scount15, "$" + scount20, "$" +  scountCu]
  return finalWord;
}

console.log(calculateTips("$10.00", "25%"))

//?Codice scritto da ChatGpt
//* Non usata 

//* ✅ PRO

//* Parsing corretto di $ e %

//* Calcoli matematici giusti

//* toFixed(2) usato bene

//* Output nel formato richiesto

//* Codice chiaro e leggibile

//! ❌ CONTRO

//! Variabili con nome sbagliato (scount → tip)

//! toString() inutile (toFixed restituisce già stringa)

//! Codice ripetuto (15%, 20%, custom)

//! Manca gestione input non valido (ok per la traccia, ma limite reale)

//? Versione Chat Gpt 

//todo Considerazioni: 

//! N.B.

File: Python/3.Ottobre/71.Tip_calculator-20Ottobre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Dati il prezzo di un pasto e una percentuale di mancia personalizzata,
# restituire un array contenente gli importi della mancia al:
# - 15%
# - 20%
# - percentuale personalizzata
#
# Il prezzo è nel formato "$N.NN".
# La percentuale è nel formato "25%".
# Restituire gli importi nel formato "$N.NN", arrotondati a due decimali.

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

def calculate_tips(meal_price, custom_tip):
    tip_conv = int(custom_tip.replace("%", ""))
    price_conv = float(meal_price.replace("$",""))

    return [
        "${:.2f}".format(price_conv * 15 / 100),
        "${:.2f}".format(price_conv * 20 / 100),
        "${:.2f}".format(price_conv * tip_conv / 100)
    ]

print(calculate_tips("$10.00", "25%"))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

def calculate_tips(meal_price, custom_tip):
    price = float(meal_price.removeprefix("$"))
    custom = int(custom_tip.removesuffix("%"))

    return [
        f"${price * percentage / 100:.2f}"
        for percentage in (15, 20, custom)
    ]

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Conversione corretta di prezzo e percentuale.
# - Formattazione a due decimali corretta.
# - Restituisce i valori nell'ordine richiesto.
# - Codice semplice e facile da seguire.

# //! Contro
# - Ripeti tre volte lo stesso calcolo cambiando solo la percentuale.
# - I nomi tip_conv e price_conv sono comprensibili, ma price e custom risultano più puliti.
# - format() funziona bene, ma le f-string sono oggi lo standard in Python.
# - Se in futuro dovessero cambiare le percentuali, dovresti modificare più righe.

# =========================================================
# //* Consigli
# =========================================================

# - Quando noti lo stesso codice ripetuto più volte, prova a usare un ciclo o una list comprehension.
# - Preferisci le f-string alla funzione format() per una sintassi più moderna e leggibile.
# - Dai nomi alle variabili che descrivano il valore, non l'operazione fatta (es. price invece di price_conv).
# - Continua a prestare attenzione alla formattazione dell'output: è uno dei tuoi punti di forza.